File: codes/ReportePeso.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ReportePeso
{
    class Program
    {
        static void Build(int node, int inicio, int fin, int[] array, int[] st)
        {
            if (inicio == fin)
                st[node] = array[inicio];
            else
            {
                int medio = (inicio + fin) / 2;

                Build(2 * node, inicio, medio, array, st);
                Build(2 * node + 1, medio + 1, fin, array, st);

                st[node] = Math.Min(st[2 * node], st[2 * node + 1]);
            }
        }

        static void Update(int node, int inicio, int fin, int i, int valor, int[] array, int[] st)
        {
            if (inicio == fin && inicio == i)
            {
                st[node] = valor;
                array[inicio] = valor;
            }
            else
            {
                int medio = (inicio + fin) / 2;

                if (i <= medio)
                    Update(2 * node, inicio, medio, i, valor, array, st);
                else
                    Update(2 * node + 1, medio + 1, fin, i, valor, array, st);

                st[node] = Math.Min(st[2 * node], st[2 * node + 1]);
            }
        }

        static int min = int.MaxValue;

        static int Query(int node, int inicio, int fin, int i, int j, int[] st)
        {
            if (fin < i || inicio > j)
                return min;
            if (inicio >= i && fin <= j)
                return st[node];

            int medio = (inicio + fin) / 2;

            int l = Query(2 * node, inicio, medio, i, j, st);
            int r = Query(2 * node + 1, medio + 1, fin, i, j, st);

            return Math.Min(l, r);
        }


        static void Main(string[] args)
        {
            int[] st = new int[4 * 100005];
            int[] array = new int[100005];

            int n = int.Parse(Console.ReadLine());

            string[] s = Console.ReadLine().Split();
            for (int i = 0; i < s.Length; i++)
            {
                array[i] = int.Parse(s[i]);
            }

            Build(1, 0, n - 1, array, st);
            int q = int.Parse(Console.ReadLine());

            while (q > 0)
            {
                string[] z = Console.ReadLine().Split();
                int b = int.Parse(z[0]);
                int c = int.Parse(z[1]);

                Update(1, 0, n - 1, b - 1, c, array, st);
                Console.WriteLine(Query(1, 0, n - 1, 0, n - 1, st));
                q--;
            }
        }
    }
}

File: codes/ReportePeso.cpp
#include <bits/stdc++.h>

using namespace std;

int arr[100005];
int st[4 * 100005];
int n, update, indice, valor, a;
int nulo = INT_MAX;

int Merge(int a, int b)
{
    return min(a, b);
}

void Build(int node, int inicio, int fin)
{
    if(inicio == fin)
        st[node] = arr[inicio];
    else
    {
        int medio = (inicio + fin) / 2;

        Build(node * 2, inicio, medio);
        Build(node * 2 + 1, medio + 1, fin);

        st[node] = Merge(st[2 * node], st[2 * node + 1]);
    }
}

void Update(int node, int inicio, int fin, int indice, int valor)
{
    if(inicio == fin && inicio == indice)
    {
        st[node] = valor;
        arr[inicio] = valor;
    }
    else
    {
        int medio = (inicio + fin) / 2;

        if (indice <= medio)
            Update(2 * node, inicio, medio, indice, valor);
        else
            Update(2 * node + 1, medio + 1, fin, indice, valor);

        st[node] = Merge(st[2 * node], st[2 * node + 1]);
    }
}

int Query(int node, int inicio, int fin, int i, int j)
{
    if (fin < i || inicio > j)
        return nulo;
    if (inicio >= i && fin <= j)
        return st[node];

    int medio = (inicio + fin) / 2;

    int l = Query(2 * node, inicio, medio, i, j);
    int r = Query(2 * node + 1, medio + 1, fin, i, j);

    return Merge(l, r);
}

int main()
{
    cin >> n;

    for(int i = 0; i < n; i++)
    {
        cin >> a;
        arr[i] = a;
    }

    Build(1, 0, n - 1);
    cin >> update;

    while(update--)
    {
        cin >> indice >> valor;
        Update(1, 0, n - 1, indice - 1, valor);
        cout << Query(1, 0, n - 1, 0, n - 1) << endl;
    }
}
